### DISCUSSION TITLE
Feature request: C# 7.3 enum generic constraint not being able to cast to integer

### DISCUSSION THREAD

**Discussion by alexdrl** [dotnet/csharplang#1993] [Q&A] [answered]
I am not being able to cast a generic method restricted type parameter to work like I supposed it worked.

```cs
public void DoSomething<T>(List<T> list)
    where T : Enum
{
    T first = list.FirstOrDefault();
    int value = (int) first;
}
```

I am missing something?

**UPDATE:**
What about something like:
```cs
public void DoSomething<T>(List<T> list)
    where T : Enum, int
{
    T first = list.FirstOrDefault();
    int value = (int) first;
}
```

**Comment by Joe4evr** [comment]
It's not guaranteed the enum backing field is an `int`. What are you gonna do if it's a `ulong`, for example?

**Comment by alexdrl** [comment]
I thought about that, what about something like this:
```cs
public void DoSomething<T>(List<T> list)
where T : Enum, int
{
T first = list.FirstOrDefault();
int value = (int) first;
}
```

**Comment by YairHalberstadt** [comment]
You could force it with dynamic:

```csharp
public void DoSomething<T>(List<T> list)
where T : Enum
{
T first = list.FirstOrDefault();
int value = (dynamic) first;
}
```

Which actually gets compiled to:

```csharp
public void DoSomething<T>(List<T> list)
where T : Enum
{
T first = list.FirstOrDefault();
int value = (int) (object) first;
}
```

**Comment by alexdrl** [comment]
In my code I solved it casting the Enum as an object and to int... I suppose that this could incur in object casting and performance cost. Will dynamic be more performant? 

Thanks for the suggestion, but not having to do this tricks would be ideal.

**Comment by YairHalberstadt** [comment]
It gets compiled to identical code in this case. The real cost here is of course boxing and unbocing, not casting which is essentially free.

If this is really performance critical you could use unsafe code, or even call System.Runtime.CompilerServices.Unsafe.As<TFrom, TTo>, which would avoid boxing

**Comment by mikernet** [comment]
I recall reading that in .NET 4.something the boxing/unboxing gets optimized out by the JIT in generic methods.

**Comment by mikernet** [comment]
Nevermind, scratch that. Possible future optimization? 

For this method:

```c#
int GetValue1<T>(T value)
{
    return (int)(object)value;
}
```

The JIT is smart enough to optimize out the box/unbox if `T` is an `int` because the types match exactly, but unfortunately it isn't smart enough to do it if `T` is a compatible enum type :-1:

**Comment by ufcpp** [comment, marked as answer]
```cs
using System;
using System.Runtime.CompilerServices;

public static class EnumExtension
{
    // size-specific version
    public static TInt AsInteger<TEnum, TInt>(this TEnum enumValue)
        where TEnum : unmanaged, Enum
        where TInt : unmanaged
    {
        if (Unsafe.SizeOf<TEnum>() != Unsafe.SizeOf<TInt>()) throw new Exception("type mismatch");
        TInt value = Unsafe.As<TEnum, TInt>(ref enumValue);
        return value;
    }

    // long version
    public static long AsInteger<TEnum>(this TEnum enumValue)
        where TEnum : unmanaged, Enum
    {
        long value;
        if (Unsafe.SizeOf<TEnum>() != Unsafe.SizeOf<byte>()) value = Unsafe.As<TEnum, byte>(ref enumValue);
        else if (Unsafe.SizeOf<TEnum>() != Unsafe.SizeOf<short>()) value = Unsafe.As<TEnum, short>(ref enumValue);
        else if (Unsafe.SizeOf<TEnum>() != Unsafe.SizeOf<int>()) value = Unsafe.As<TEnum, int>(ref enumValue);
        else if (Unsafe.SizeOf<TEnum>() != Unsafe.SizeOf<long>()) value = Unsafe.As<TEnum, long>(ref enumValue);
        else throw new Exception("type mismatch");
        return value;
    }
}
```

**Comment by yaakov-h** [comment]
@ufcpp how about unsigned? 😄

**Comment by YairHalberstadt** [comment]
@ufcpp
Shouldn't some of those `!=` be `==`?

**Comment by ufcpp** [comment]
@yaakov-h 
This checks only size of the `TEnum`, thus `byte` and `sbyte` are the same.

@YairHalberstadt 
Mistake. `==` is correct.

**Comment by mikernet** [comment]
> It gets compiled to identical code in this case.

@YairHalberstadt Wait a sec...it definitely does not. That dynamic cast can't possibly compile to the identical code above. Written the way you have it:

```c#
int value = (dynamic) first;
```

That line of code needs to look for an implicit conversion between whatever type `first` is (and to `dynamic` it's just an `object` until it calls `GetType()`) to `int`.

Even if you add an explicit cast there's still no way it emits the same code. The explicit dynamic cast will work fine for something like a `byte` sized enum being cast to an `int`, but casting a `byte` sized enum stored in an `object` to `int` definitely doesn't work.

This all involves dynamic call site binding if you start using `dynamic`.

**Comment by YairHalberstadt** [comment]
@mikernet
Take a look at sharplab

**Comment by mikernet** [comment]
@YairHalberstadt Look at the IL view in SharpLab. That's not the code that gets emitted. It's decompiler either has a bug or intentionally changes the code, but that's not what actually gets emitted and the performance of the emitted code isn't the same as it's decompiled view.

**Comment by mikernet** [comment]
Your dynamic method doesn't even work - it throws a `RuntimeBinderException`:  Cannot implicitly convert type 'UserQuery.Test' to 'int'. An explicit conversion exists (are you missing a cast?)

(`UserQuery.Test` is an `int` backed enum)

**Comment by mikernet** [comment]
Somewhat shockingly, the dynamic cast performs faster than the normal cast even though it is still going through the boxing process. I can't believe that this:

```
IL_0000:  ldsfld      UserQuery+<>o__3<>.<>p__0
IL_0005:  brtrue.s    IL_002C
IL_0007:  ldc.i4.s    10 
IL_0009:  ldtoken     System.Int32
IL_000E:  call        System.Type.GetTypeFromHandle
IL_0013:  ldtoken     UserQuery
IL_0018:  call        System.Type.GetTypeFromHandle
IL_001D:  call        Microsoft.CSharp.RuntimeBinder.Binder.Convert
IL_0022:  call        System.Runtime.CompilerServices.CallSite<System.Func<System.Runtime.CompilerServices.CallSite,System.Object,System.Int32>>.Create
IL_0027:  stsfld      UserQuery+<>o__3<>.<>p__0
IL_002C:  ldsfld      UserQuery+<>o__3<>.<>p__0
IL_0031:  ldfld       System.Runtime.CompilerServices.CallSite<System.Func<System.Runtime.CompilerServices.CallSite,System.Object,System.Int32>>.Target
IL_0036:  ldsfld      UserQuery+<>o__3<>.<>p__0
IL_003B:  ldarg.1     
IL_003C:  ldc.i4.0    
IL_003D:  callvirt    System.Collections.Generic.List<>.get_Item
IL_0042:  box         UserQuery.T
IL_0047:  callvirt    System.Func<System.Runtime.CompilerServices.CallSite,System.Object,System.Int32>.Invoke
IL_004C:  pop         
IL_004D:  ret
```

Performs better than this:

```
IL_0000:  ldarg.1     
IL_0001:  ldc.i4.0    
IL_0002:  callvirt    System.Collections.Generic.List<>.get_Item
IL_0007:  box         UserQuery.T
IL_000C:  unbox.any   System.Int32
IL_0011:  pop         
IL_0012:  ret 
```

That's some incredible DLR voodoo magic going on. The dynamic cast does behave differently so maybe its higher speed can be attributed to less type checks going on or something.

Neither performs nearly as well as the Unsafe method @ufcpp posted above or [my Caster implementation here](https://gist.github.com/mikernet/c68c1251d6a7e4018cee85783528b589). I was curious how the speed of the `Unsafe.As` approach compares to my compiled expression approach so I tested it and it's almost identical.

`Caster` can be used for casting between any two types without any unsafeness or matching size requirements and supports checked context casts - in case someone reading this finds it useful here are some typical usages (first method is the one I performance tested against `Unsafe.As`):

```c#
int value = Caster.Cast<EnumType, int>(EnumType.SomeValue);
int value = Caster.CheckedCast<long, int>(someLongValue);

object intValue = (int)5;
object ulongValue = intValue.DynamicCastTo(typeof(ulong), checkOverflow: true);
```
Everything has been meticulously tuned to give the best possible performance in every situation given the information you provide the methods.

**Comment by jnm2** [comment]
@mikernet 

>  Look at the IL view in SharpLab. That's not the code that gets emitted. It's decompiler either has a bug or intentionally changes the code, but that's not what actually gets emitted and the performance of the emitted code isn't the same as it's decompiled view.

I've used it a lot and have not noticed anything amiss, but if this is the case, Andrey will surely want to know about it. What version of the compiler do you have selected in SharpLab, and do you have Debug or Release selected?

**Comment by mikernet** [comment]
@jnm2 If you look up you can see the referenced issue that @svick already created in ILSpy's decompiler repo for this :)

**Comment by ufcpp** [comment]
@mikernet
`Unsafe.SizeOf` is JIT-time constant and all the `if` statements are removed. Moreover, `Unsafe.As` is inlined. My Unsafe method is almost as fast as non-generic implmentation.

**Comment by mburbea** [comment]
Have you considered `GetHashCode`? Which while kind of a strange workaround guarantees to get back an int, even if the backing type is not an integer.
According to this issue, it doesn't box:: https://github.com/dotnet/coreclr/issues/8143

**Comment by svick** [comment]
@mburbea `GetHashCode` is guaranteed to give you an `int`, but there is no guarantee about *which `int`* it will give you, so I think using it here is a bad idea.

**Comment by mikernet** [comment]
@ufcpp Yes I know. The only additional overhead my method has is a generic cached delegate invocation which is why it runs at almost the same speed. 10 million conversions using your method took 19ms, mine took 27ms. I know `Unsafe.As` is fast, that's why I wanted to see how mine compared to it and I'm rather happy with the results, can't do much better than that :smile:

**Comment by mburbea** [comment]
@svick, sure, but if you're blindly assuming that the type is an int. Are you really concerned? If the type isn't long or ulong, you'll get the desired result.

This is still safer than the suggested `Unsafe.As`.

**Comment by mikernet** [comment]
@mburbea Just because `Unsafe.As` was used doesn't mean it it's actually unsafe. The way it's written looks pretty safe to me. Care to point out a situation where it wouldn't be?

There's always my solution if you don't like unsafe code :)

**Comment by mburbea** [comment]
Sorry, I retract my objection. That `Unsafe.As` is perfectly fine. I was assuming it was just blindly using `UnsafeAs`, but that solution is safe and should produce good quality code.

**Comment by Ultrahead** [comment]
@alexdrl: maybe you could define a converter (?):

```csharp
        public static class EnumConverter<TInput, TOutput> 
            where TInput: struct, Enum
            where TOutput : struct, IConvertible
        {
            public static readonly Func<TInput, TOutput> Convert = GenerateConverter();

            private static Func<TInput, TOutput> GenerateConverter()
            {
                var parameter = Expression.Parameter(typeof(TInput), typeof(TInput).Name);
                var dynamicMethod = Expression.Lambda<Func<TInput, TOutput>>
                (
                    Expression.Convert(parameter, typeof(TOutput)),
                    parameter
                );
                return dynamicMethod.Compile();
            }
        }
```

And then you could call it either directly or define an extension method for the specific enum type you want to cast to int:

```csharp
        public static int ToInt(this MySpecificEnum source)
        {
            return EnumConverter<MySpecificEnum, int>.Convert(source);
        }
```

**Reply by mariangemarcano** [reply]
Another option based on @Ultrahead  suggestion, limited to `int` and adding some extra checks to prevent silent errors, making sure that the type does not convert to the wrong value, and throwing an exception if the value is not defined by enum.


```
  public static class EnumConverter<TInput>
        where TInput : struct, Enum
    {
        private static readonly Func<TInput, int> _convertFrom = GenerateConverter<TInput, int>();
        private static readonly Func<int, TInput> _convertTo = GenerateConverter<int, TInput>();

        private static Func<TInput, TOutput> GenerateConverter<TInput, TOutput>()
        {
            var parameter = Expression.Parameter(typeof(TInput), typeof(TInput).Name);
            var dynamicMethod = Expression.Lambda<Func<TInput, TOutput>>
            (
                // Make sure the type is not silently converted to a wrong value (overflow)
                Expression.ConvertChecked(parameter, typeof(TOutput)),
                parameter
            );
            return dynamicMethod.Compile();
        }

        /// <summary> Converts a generic Enum to int </summary>
        public static int AsInt(TInput input) => _convertFrom(input);

        /// <summary> Converts from int to generic Enum type </summary>
        public static TInput AsEnum(int input)
        {
            var result = _convertTo(input);

            if (!Enum.IsDefined(typeof(TInput), result))
                throw new ArgumentOutOfRangeException("Value is not defined in the enum.");

            return result;
        }
    }
```

**Comment by alexdrl** [comment]
@Ultrahead I do not need the initial code anymore, and your code needs to know in advance that the enum is an integer.

What I initially proposed is that the compiler supported the additional compiler check that the enum passed to the generic method has an underlying specific type (int, ulong...). So that my compile Will not compile if the passed enum does not have the specfied type.

Thank you for your comments, I have learned a lot about .NET performance internals.

**Comment by Ultrahead** [comment]
Alejandro: es bueno saberlo. I posted the code as a workaround, though, given its performace (it avoids boxing).

I didn't see the proposal, besides the title it-self. Because even in the case you present, the compiler should know in advance your intent. Maybe a way to clearly state should be allowing something like:

where T: Enum, int

**Comment by alexdrl** [comment]
>I didn't see the proposal, besides the title it-self. Because even in the case you present, the compiler should know in advance your intent. Maybe a way to clearly state should be allowing something like:
>
>where T: Enum, int

@Ultrahead Yes, If the compiler knows the underlying type, why not having the possibility to use it as a generic method constraint? Or the compile does not know the underlying type for an enum?

Maybe I should rephrase the issue title.

**Comment by Ultrahead** [comment]
It does know but it's a matter of specifying your intent in the signature to guarantee and enforce compilation. That's why I mentioned:

where T: Enum, int

or even T: Enum(int)

The title is ok.

**Comment by mikernet** [comment]
There are probably a couple places I would use this if it was available since a compile time error is much nicer than a runtime error if you require an enum of a certain underlying type but that's such a niche situation that I can't imagine this will get a lot of priority.

**Comment by mikernet** [comment]
Syntax wise I would imagine it would have to be more like the second example @Ultrahead gave because it's not actually an `int` so the other way would be very misleading.

**Comment by Ultrahead** [comment]
@mikernet: yes, I realized that after posting the first syntax. I agree.
